mlir/lib/Target/LLVMIR/DebugImporter.cpp

```diff
@@ -289,7 +316,19 @@ DINodeAttr DebugImporter::translate(llvm::DINode *node) {
     return nullptr;
   };
   if (DINodeAttr attr = translateNode(node)) {
-    nodeToAttr.insert({node, attr});
+    // If this node was marked as recursive, set its recId.
+    if (auto recType = dyn_cast<DIRecursiveTypeAttrInterface>(attr)) {
+      if (DistinctAttr recId = translationStack.lookup(node)) {
+        attr = cast<DINodeAttr>(recType.withRecId(recId));
+        // Remove the unbound recursive ID from the set of unbound self
+        // references in the translation stack.
+        unboundRecursiveSelfRefs.back().erase(recId);
+      }
+    }
+
+    // Only cache fully self-contained nodes.
+    if (unboundRecursiveSelfRefs.back().empty())
```

Comment:
@zyx-billy the import now works on all our tests but we found a efficiency issue. One of our test files grows by roughly factor 10x when importing and I see may versions of the same unbounded version of a composite type. I suspect this is the case nested recursive types. Then the same DICompositeType may be translated several times with different distinct identifiers since only once we left the outermost cycle we actually cache the type.

The solution to this seems not so trivial since caching a composite type is only really possible once we know there is no unbound type. Could we maybe have some sort of temporary cache during the traversal that keeps track of translated recursive types given a set of currently open unbounded references. That could possibly handle the case where in a cycle there are several entry nodes into a nested cycle...

Didn't have time to think about this in depth yet but you may have had thoughts when implementing this.

Reply:
oh very interesting... Is it something like this
```
// LLVM
+---> B ----+
|           v
A <-------- D
|           ^
+---> C ----+

// MLIR
A -> B -> D -> A'
+--> C -> D -> A'
```
So now in MLIR you have two copies of the subgraph `D -> A'`, but they're different because we don't cache it?

I like your idea of a temporary cache that keys on the set of unbounded references. This way repeated subgraphs can be merged too. I can get started on something for this. Please let me know if this is not what you were looking for though 😂.

Reply:
Oh I suppose in the above case, D also has to contain another cycle (perhaps D <-> E), this way D is not simply deduplicated by MLIR. So something like
```
// LLVM
+---> B ----+
|           v
A <-------- D <--> E
|           ^
+---> C ----+

// MLIR
A -> B -> D0 -> A'
|         + --> E0 --> D0'
|
+--> C -> D1 -> A'
          + --> E1 --> D1'
```
D0 & D1 cannot be cached because they contain an unbound A', but also are not identical because they have distinct recIds.

Reply:
Yes I was thinking of the last example you posted with the nested cycles. I am not 100% sure if this is what we are seeing though our input is 1.6gb of ir...

What I tried though is the following hack:
```
      if (DistinctAttr recId = translationStack.lookup(node)) {
        attr = cast<DINodeAttr>(recType.withRecId(recId));
        // Remove the unbound recursive ID from the set of unbound self
        // references in the translation stack.
        unboundRecursiveSelfRefs.back().erase(recId);

        if (!unboundRecursiveSelfRefs.back().empty()) {
          DINodeAttr containedAttr = nullptr;
          for (auto &[unboundSelfRefs, attr] : unboundAttrCache[node]) {
            if (unboundSelfRefs == unboundRecursiveSelfRefs.back())
              containedAttr = attr;
          }
          if (!containedAttr)
            unboundAttrCache[node].emplace_back(unboundRecursiveSelfRefs.back(),
                                                attr);

          if (containedAttr) {
            attr = containedAttr;
          }
        }
      }
    }
```
So essentially I cache previous instances of a recursive attribute declaration with the corresponding unbounded self refs and then always return an attribute from the cache... This solves the IR size problem but not the compilation time problem since the traversal is unchanged. I am not really sure the IR is correct but at least it imports and exports...

Reply:
This is the minimal example I can come up with:
```
define void @class_field(ptr %arg1) !dbg !18 {
  ret void
}

declare void @llvm.dbg.value(metadata, metadata, metadata)

!llvm.dbg.cu = !{!1}
!llvm.module.flags = !{!0}
!0 = !{i32 2, !"Debug Info Version", i32 3}
!1 = distinct !DICompileUnit(language: DW_LANG_C, file: !2)
!2 = !DIFile(filename: "debug-info.ll", directory: "/")


!3 = !DICompositeType(tag: DW_TAG_class_type, name: "A", file: !2, line: 42, flags: DIFlagTypePassByReference | DIFlagNonTrivial, elements: !4)
!4 = !{!7, !8}

!5 = !DICompositeType(tag: DW_TAG_class_type, name: "B", scope: !3, file: !2, line: 42, flags: DIFlagTypePassByReference | DIFlagNonTrivial, elements: !9)
!6 = !DIDerivedType(tag: DW_TAG_pointer_type, baseType: !5, flags: DIFlagArtificial | DIFlagObjectPointer)
!7 = !DIDerivedType(tag: DW_TAG_member, name: "B:B1", file: !2, baseType: !5)
!8 = !DIDerivedType(tag: DW_TAG_member, name: "B:B2", file: !2, baseType: !5)
!9 = !{!7, !8}

!18 = distinct !DISubprogram(name: "A", scope: !3, file: !2, spFlags: DISPFlagDefinition, unit: !1)
```
When importing the B type declaration duplicates:
```
#di_composite_type3 = #llvm.di_composite_type<tag = DW_TAG_class_type, recId = distinct[1]<>, name = "B", file = #di_file, line = 42, scope = #di_composite_type, flags = "TypePassByReference|NonTrivial", elements = #di_derived_type, #di_derived_type1>
#di_composite_type4 = #llvm.di_composite_type<tag = DW_TAG_class_type, recId = distinct[2]<>, name = "B", file = #di_file, line = 42, scope = #di_composite_type, flags = "TypePassByReference|NonTrivial", elements = #di_derived_type2, #di_derived_type3>
```
Maybe it also makes sense to key the cache on the translation stack since the unboundSelfRefs are only really known after traversing the subtree if I understand correctly. Anyways I hope the example helps to clarify the problem!

Reply:
PS sorry for posting on the weekend. I just had time now to extract such an example.

Reply:
Oh thank you so much for the example. I see now that caching on the set of unbound self-refs doesn't solve the compile time problem... It probably does make sense to instead just create a new cache for each level of the translation stack like you said. Every time we put something that is potentially recursive onto the stack, we start a new cache. And one trick we can play is every time we finish a recursive type, we go into the current layer cache, replace all instances of its rec-self with the just-finished recursive type, and insert into the outer layer cache. This way `translate` never walks into a recursive type twice. Let me try to make something like this that's efficient 🤔

Reply:
>  And one trick we can play is every time we finish a recursive type, we go into the current layer cache, replace all instances of its rec-self with the just-finished recursive type, and insert into the outer layer cache.

Ah right, with that we should be on the safe side.

Reply:
@gysit I have a solution here that should work: https://github.com/llvm/llvm-project/pull/87295. For your example, it only creates one copy of the B type declaration.
In addition to caching every translated result, the unbound self-ref replacement is lazy so that it's only triggered on use. Interested to see how this works for your cases. Hoping that both compile time and result IR is smaller 🙏. I can open it up for review if it works, otherwise we can see where to further optimize. Thanks!

Reply:
Very nice!

I see a massive reduction in size and it is comparable to llvm now:
```
-rw-r--r-- 1 tobiasg tobiasg  196553791 Apr  2 10:59 sparta-input.ll
-rw-r--r-- 1 tobiasg tobiasg  162484418 Apr  2 10:47 sparta-with-fix.mlir
-rw-r--r-- 1 tobiasg tobiasg 1817219645 Apr  2 10:59 sparta-without-fix.mlir
```
Additionally compilation time drops by factor 10x as well. I have a debug build only though. However, I expect this reproduce in release build.

So I think we should probably go ahead with your solution!